[Charles Lim Jun Hao] iP

charleslimjh opened the pull request:
# Duke

> DukeBot is a **CLI** application that helps *manage your tasks*. 👍 

You can:
- Add tasks
- Delete tasks
- Mark tasks as complete/incomplete

To get started:
1. Download the `.jar` file from [here](https://github.com/charleslimjh/ip/releases/tag/A-Jar).
2. Navigate to the file in your terminal.
3. Run the following command `java -jar duke.jar`
4. Profit!

Features to implement:
- [x] Basic CRUD
- [x] Load to file
- [ ] GUI
- [ ] Testing 

Here's a preview of the main code:
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

charleslimjh committed:
implemented echo function

charleslimjh committed:
implemented add/list functions

charleslimjh committed:
implemented Task class

charleslimjh committed:
added toString for Task

charleslimjh committed:
integrated ToDo into Duke

charleslimjh committed:
modified toggleDone to setDone and setUndone

charleslimjh committed:
implemented mark/unmark functionality

charleslimjh committed:
Converted static methods to non-static

charleslimjh committed:
Implemented Todo class

charleslimjh committed:
Extracted functions out of processInput

charleslimjh committed:
Reordered functions and improved javadocs

charleslimjh committed:
Fix typo in javadocs

charleslimjh committed:
Implemented Event and Deadline classes

charleslimjh committed:
fixed typo in toString

charleslimjh committed:
Integrated event/ deadline/ to-do to Duke

charleslimjh committed:
bug fixes for adding tasks

charleslimjh committed:
simplified Duke's greeting

charleslimjh committed:
implemented automated text ui testing

charleslimjh committed:
add DukeException class

charleslimjh committed:
handle invalid command and missing task description errors

charleslimjh committed:
create directory and file

charleslimjh committed:
implement toStringWritable for Task classes

charleslimjh committed:
save tasks after each command

charleslimjh committed:
overload Task constructors to read from file

charleslimjh committed:
read from file

charleslimjh committed:
integrate LocalDateTime into event and deadline

charleslimjh committed:
exception handling in duke

charleslimjh committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
load dates from file

sohenze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,247 @@
+        boolean goodbye = false;
+
+        try {
+            if (s.equals("bye")) {
+                System.out.println("Bye. Hope to see you again!");
+                goodbye = true;
+                return true;
+            } else if (s.equals("list")) {
+                this.list();
+            } else if (userInput[0].equals("todo")) {
+                this.addTodo(userInput);
+            } else if (userInput[0].equals("deadline")) {
+                this.addDeadline(userInput);
+            } else if (userInput[0].equals("event")) {
+                this.addEvent(userInput);
+            } else if (userInput[0].equals("mark")) {
+                this.mark(userInput);
+            } else if (userInput[0].equals("unmark")) {
+                this.unmark(userInput);
+            } else if (userInput[0].equals("delete")) {
+                this.delete(userInput);
+            } else {
+                throw new DukeException("Unrecognized command.");
+            }
```
Perhaps a switch statement would be better here?

charleslimjh replied:
Yup, just implemented it. Good suggestion!

sohenze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,247 @@
+import java.util.stream.Stream;
+
 public class Duke {
+    private ArrayList<Task> db = new ArrayList<>(10);
```
Perhaps a more intuitive variable name here?

charleslimjh replied:
Personally, I thought `db` was a very intuitive name since it commonly refers to 'database'. A good point to consider though, thank you!

sohenze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,247 @@
+                tmp[i] = tmp[i].strip();
+            }
+
+            switch (tmp[0]) {
+                case "T":
+                    db.add(new Todo(tmp));
+                    break;
+                case "D":
+                    db.add(new Deadline(tmp));
+                    break;
+                case "E":
+                    db.add(new Event(tmp));
+                    break;
```
I like how you used a switch statement here! But take note that we are not supposed to have indentations for case clauses!

charleslimjh committed:
implementing more OOP in Duke

charleslimjh committed:
implement Ui class

charleslimjh committed:
implement Storage class

charleslimjh committed:
implement abstract Command class

charleslimjh committed:
implement TaskList class

charleslimjh committed:
implement bye and list commands

charleslimjh committed:
fix typo in addDeadline

charleslimjh committed:
implement Todo/Event/Deadline Commands

charleslimjh committed:
remove unused import statements

charleslimjh committed:
overload constructors for event/deadline

charleslimjh committed:
remove print statements in TaskList

charleslimjh committed:
remove extra print statement in Ui

charleslimjh committed:
implement mark/unmark commands

charleslimjh committed:
implement Parser

charleslimjh committed:
implement delete Command

charleslimjh committed:
write changes to file in mark/unmark/delete commands

charleslimjh committed:
print tab before tasks in delete/mark/unmark commands

charleslimjh committed:
organize into packages

charleslimjh committed:
Merge remote-tracking branch 'ip2/add-gradle-support'

charleslimjh committed:
fix typo in index out of bounds exception

charleslimjh committed:
configure gradle to build and run

charleslimjh committed:
create JUnit test for TaskList

charleslimjh committed:
configure build.gradle to create JAR files

charleslimjh committed:
add javadocs to main duke classes

charleslimjh committed:
add javadocs to command classes

charleslimjh committed:
add javadocs to task classes

charleslimjh committed:
add javadocs to DukeException class

charleslimjh committed:
implement find functionality

charleslimjh committed:
Merge branch 'branch-Level-9'

charleslimjh committed:
implement basic GUI

charleslimjh committed:
implement getResponse in Duke to take in commands and return response to GUI

charleslimjh committed:
add getResponse to all Commands to return strings to GUI

charleslimjh committed:
implement GUI using FXML

charleslimjh committed:
Fix bug in build.gradle

mainClassName was pointing to the wrong location, so gradle could not
start Duke properly.

Fixing the typo now allows gradle to identify the correct file and to
run the application properly.

charleslimjh committed:
Include assert statements in Parser and TaskList

charleslimjh committed:
Delete dead code

After implementation of GUI, no need to rely on text-interface commands.

Let's,
* remove Duke : run()
* remove Command : execute()

charleslimjh committed:
Refactor Parser logic

Previous parse command was too long and did not conform to SLAP.
Nested conditional statements were also present.

Let's,
* Abstract out how to process each input command
* Utilize more guard clauses in favor of nested conditionals

charleslimjh committed:
Refactor TaskList magic numbers

TaskList utilizes a 0-indexed ArrayList, while Tasks are 1-indexed
according to the GUI. Code does not clearly reflect this.

Let's refactor out the (-1) and (+1) to a named constant for clarity.

charleslimjh committed:
Merge pull request #2 from charleslimjh/branch-A-Assertions

Add assertions

charleslimjh committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/TaskList.java

charleslimjh committed:
Merge pull request #3 from charleslimjh/branch-A-CodeQuality

Improve code quality

charleslimjh committed:
Implement simple help function

charleslimjh committed:
Tweak help function to redirect to GitHub User Guide

charleslimjh committed:
Add user guide for Duke

charleslimjh committed:
Update README.md

Fix table formatting and included links

charleslimjh committed:
Add additional help text

charleslimjh committed:
Make GUI responsive to resizing and text overflow

charleslimjh committed:
Merge remote-tracking branch 'origin/master'
